docs/providers/aws/guide/credentials.md

```diff
@@ -210,3 +210,7 @@ custom:
 Thus, most likely, a different set of permissions will be in place, altering the interaction between your lambda functions and other AWS resources.
 
 _Please, refer to the [`invoke local`](https://serverless.com/framework/docs/providers/aws/cli-reference/invoke-local/#aws---invoke-local) CLI command documentation for more details._
+
+#### EC2 IAM Role
+
+If you run serveless inside an Amazon EC2 instance, you can leverage IAM roles for Amazon EC2 to automatically provide credentials to it. If you configure your instance to use IAM roles, the AWS SDK automatically selects the IAM credentials for your serverless, eliminating the need to manually provide credentials. If you run any command inside of an EC2 instance and are leveraging IAM roles, you should consider using `--ignore-metadataservice-validation` option in CLI command, Eg.: `serverless deploy --ignore-metadataservice-validation`
```

Comment:
To make sure I understand right.

Our current check (as in master) is technically good, but immune to timeouts?
So this option is added purely to avoid pursuing that check at all, if someone observes the timeouts. Right?

Reply:
If it's the case, wouldn't bumping timeout to 1s, fix it for most cases?

Reply:
So, bumping the timeout is an option, however it is not guaranteed that it won't fail intermittently, and if you are in a CI/CD process, you don't want your deployment to fail intermittently.

Reply:
Ok, but if even with increased timeout validation can occasionally fail, doesn't that mean that AWS authorization (which happens afterwards) also can occasionally fail?

Reply:
The metadata service is not responsible for authorisation, it is just an easier way to find if you are running AWS SDK inside of an EC2 machine. BTW, it is not meant for that.
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html

And for sure, any AWS SDK commands can timeout, regarding of how credentials are set, however we usually don't change the timeout when invoking AWS Commands.

I believe the SDK sets a very generous timeout when calling the AWS Endpoints, for the metadata service it doesn't as you can see here. https://github.com/aws/aws-sdk-js/blob/4ed556871d347b07e95da4ec5dd0ffe1bcafe87f/lib/metadata_service.js#L38

Reply:
Thanks @pauloprestes for clarifications. I've looked into it more closely, and to my understanding situation is as follows:

- `metadataService.request()`  doesn't confirm really on credentials, but on fact whether we're in AWS EC2 environment - So that check is ineffective, as even when succeeds it doesn't mean we're properly authenticated
- In EC2 environment it's possible to authenticate via attached through AWS console IAM role, and this setup cannot be detected via inspecting environment vars
- Environment credentials could be validated via:

```javascript
const { Config } = require('aws-sdk');
require('aws-sdk/lib/node_loader');
const config = new Config();

console.log("Is authenticated?", Boolean(config.credentials));
```

If that's correct I would apply following changes:
- Replace  https://github.com/serverless/serverless/blob/1cc8778d7f370a3bfa09ff8a858f7878649cb7b8/lib/plugins/aws/lib/validate.js#L33-L59 with check as proposed above.

What do you think?

Reply:
I've tested the checked proposed and it returns false inside of an EC2.
Would you be ok if we just increase the connection timeout to 1 second, then? I think it would work most of the times.

Reply:
Indeed, technically the only way to confirm is try to make some AWS request, even to non-existing resource.. still it's not clear to me how to configure for such case request timeout so request doesn't hang in case of missing credentials (e.g. as I tested with lambda invocation, any timeout settings doesn't seem to work).

Concerning metadata check, I think it should be ditched, as it doesn't confirm on AWS credentials, in EC2 it'll pass even with credentials not setup.

Reply:
I've updated the PR to remove completely the EC2 metadata service check.
I think this would be the best approach, to be honest, however if someone is using serverless to deploy from ec2 machine, the builds will fail and they will need to add the option, so I believe it becomes a breaking change.

Reply:
Unfortunately we cannot make breaking change without shifting major, and also we cannot at this point shift major due to our roadmap constraints.

Technically what we're trying to workaround here is the AWS SDK behavior, where it hangs on missing credentials. Real solution is to find a workaround for that. I've posted a question to AWS SDK -> https://github.com/aws/aws-sdk-js/issues/2983

Reply:
So, I agree the best solution for this behaviour, it is to change js aws sdk to timeout the request.
Go library already does that and it adds a timeout of 5 seconds.
https://github.com/aws/aws-sdk-go/blob/722a38b2f1be499ffd31d849e0a2e4acf17782e3/aws/ec2metadata/service.go#L83

Also, I found a method which retrieves the credentials for the ec2 machine. https://github.com/aws/aws-sdk-js/blob/4ed556871d347b07e95da4ec5dd0ffe1bcafe87f/lib/metadata_service.js#L135

And I found some comments suggesting that 5 seconds timeout when using it, on js library: https://github.com/aws/aws-sdk-js/blob/2a872adac4f1c49d45feb5b368e72adf0c343cf4/lib/credentials/ec2_metadata_credentials.js#L19

So, i've changed this PR to use the `loadCredentials` and increased the timeout to 5 seconds.

Reply:
Hey @jchris thanks a lot for looking into this and sharing the links :+1:

Given the links above I'd also say that this totally makes sense. The only concern I have is with the `loadCredentials` method we're using. According to the doc comments it's meant to be `private` so it might change in the future and could be unreliable as AWS doesn't expose it "publicly".

https://github.com/aws/aws-sdk-js/blob/4ed556871d347b07e95da4ec5dd0ffe1bcafe87f/lib/metadata_service.js#L175

Is there any chance to not use `loadCredentials`? Here are the publicly available methods: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/MetadataService.html

Reply:
Thanks for the finding @pmuens. Based on the links that I've sent above, it is easier to use `EC2MetadataCredentials` that is is also loading the credentials. I've updated the PR to use it.

Reply:
Thanks for updating @pauloprestes :+1: That makes sense.

I'll leave it to @medikoo to make the final call here since he's more familiar with the underlying issue(s).